File: leetcode/queue_stack/3221_Maximum_Array_Hopping_Score_II/codes/arr.c
#include <stdio.h>

int maxScore(int* nums, int n) {
    int max_v = nums[n - 1];
    int max_i = n - 1;

    int idx[n];
    for (int i = n - 2; i >= 0; --i) {
        idx[i] = max_i;
        if (nums[i] > max_v) {
            max_v = nums[i];
            max_i = i;
        }
    }

    int score = 0, i = 0;
    while (i != n - 1) {
        int j = idx[i];
        score += (j - i) * nums[j];
        i = j;
    }
    return score;
}

File: leetcode/queue_stack/3221_Maximum_Array_Hopping_Score_II/codes/arr.py
from typing import List

class Solution:
    def maxScore(self, nums: List[int]) -> int:
        n = len(nums)
        max_v = nums[-1]
        max_i = n - 1

        idx = [0] * n
        for i in range(n - 2, -1, -1):
            idx[i] = max_i
            if nums[i] > max_v:
                max_v = nums[i]
                max_i = i

        score, i = 0, 0
        while i != n - 1:
            j = idx[i]
            score += (j - i) * nums[j]
            i = j
        return score
